envoy/common/inline_map_registry.h

```diff
@@ -0,0 +1,513 @@
+};
+
+class InlineMapRegistryManager {
+public:
+  static const std::vector<InlineMapRegistryDebugInfo>& registryInfos() {
+    // Call finalize() to ensure that all finalizers are called.
+    finalize();
+    return mutableRegistryInfos();
+  }
+
+  /**
+   * Generate a scope id for the given scope and register a finalizer to be called when
+   * InlineMapRegistry::finalize() is called.
+   *
+   * NOTE: This is called by InlineMapRegistry::scopeId() and should never be called
+   * manually.
+   */
+  template <class Scope> static uint64_t generateScopeId();
+
+private:
+  using Finalizer = std::function<InlineMapRegistryDebugInfo()>;
+
+  static uint64_t nextScopeId() {
+    static uint64_t next_scope_id = 0;
+    return next_scope_id++;
+  }
+
+  /**
+   * Call all finalizers. This should only be called once, after all registrations are
+   * complete.
+   */
+  static void finalize() {
+    while (!mutableFinalizers().empty()) {
+      mutableRegistryInfos().push_back(mutableFinalizers().back()());
+      mutableFinalizers().pop_back();
+    }
+  }
+
+  static std::vector<InlineMapRegistryDebugInfo>& mutableRegistryInfos() {
+    MUTABLE_CONSTRUCT_ON_FIRST_USE(std::vector<InlineMapRegistryDebugInfo>);
+  }
+
+  static std::vector<Finalizer>& mutableFinalizers() {
+    MUTABLE_CONSTRUCT_ON_FIRST_USE(std::vector<Finalizer>);
+  }
+};
+
+/**
+ * This template could be used as an alternative of normal hash map to store the key/value pairs.
+ * But by this template, you can register some frequently used keys as inline keys and get the
+ * handle for the key. Then you can use the handle to access the key/value pair in the map without
+ * even one time hash searching.
+ *
+ * This is useful when some keys are frequently used and the keys are known at compile time or
+ * bootstrap time. You can get superior performance by using the inline handle. For example, the
+ * filter state always uses the filter name as the key and the filter name is known at compile time.
+ * By using the inline handle, the filter state could get the key/value pair without any hash.
+ *
+ * This template also provides the interface to access the key/value pair by the normal key and the
+ * interface has similar searching performance as the normal hash map. But the insertion and removal
+ * by the normal key is slower than the normal hash map.
+ *
+ * The Scope template parameter is used to distinguish different inline map registry. Different
+ * Scope will have different inline key registrations and different scope id.
+ *
+ * These is usage example:
+ *
+ *   // Define a scope type.
+ *   class ExampleScope {
+ *   public:
+ *     static absl::string_view name() { return "ExampleScope"; }
+ *   };
+ *
+ *   // Register the inline key. We should never do this after bootstrapping. Typically, we can
+ *   // do this by define a global variable.
+ *   auto inline_handle = InlineMapRegistry<ExampleScope>::registerInlineKey("inline_key");
```

Comment:
inline_handle then becomes a static-init fiasco. But if you make this library's API thread compatible like STL, this will not be an issue.

Again it's fine to define a registry class here, it is fundamental to the usage model. Just let the caller decide if they want a static one of those or not, and worry about init-order.